File: src/Stronghold.Application/Features/Gamification/LeaderboardResponse.cs
namespace Stronghold.Application.Features.Gamification;

public class LeaderboardResponse
{
    public int Rank { get; set; }
    public int UserId { get; set; }
    public string FullName { get; set; } = string.Empty;
    public string Username { get; set; } = string.Empty;
    public string? ProfileImageUrl { get; set; }
    public int XP { get; set; }
    public int Level { get; set; }
    public string LevelName { get; set; } = string.Empty;
    public int TotalGymMinutes { get; set; }
}

File: stronghold_desktop/lib/features/users/models/leaderboard_response.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'leaderboard_response.freezed.dart';
part 'leaderboard_response.g.dart';

@freezed
abstract class LeaderboardEntry with _$LeaderboardEntry {
  const factory LeaderboardEntry({
    required int rank,
    required int userId,
    required String fullName,
    required String username,
    String? profileImageUrl,
    required int xp,
    required int level,
    required String levelName,
    required int totalGymMinutes,
  }) = _LeaderboardEntry;

  factory LeaderboardEntry.fromJson(Map<String, dynamic> json) =>
      _$LeaderboardEntryFromJson(json);
}

@freezed
abstract class PagedLeaderboardResponse with _$PagedLeaderboardResponse {
  const factory PagedLeaderboardResponse({
    required List<LeaderboardEntry> items,
    required int totalCount,
    required int totalPages,
    required int currentPage,
    required int pageSize,
  }) = _PagedLeaderboardResponse;

  factory PagedLeaderboardResponse.fromJson(Map<String, dynamic> json) =>
      _$PagedLeaderboardResponseFromJson(json);
}
